editor/src/core/model/get-uid-mappings.ts

```diff
@@ -0,0 +1,273 @@
+      extractUidFromJSXElementChild(
+        working,
+        workingDupliactedUIDs,
+        filePath,
+        topLevelElement.rootElement,
+      )
+      if (topLevelElement.arbitraryJSBlock != null) {
+        extractUIDFromArbitraryBlock(
+          working,
+          workingDupliactedUIDs,
+          filePath,
+          topLevelElement.arbitraryJSBlock,
+        )
+      }
+      break
+    case 'ARBITRARY_JS_BLOCK':
+      extractUIDFromArbitraryBlock(working, workingDupliactedUIDs, filePath, topLevelElement)
+      break
+    case 'IMPORT_STATEMENT':
+      break
+    case 'UNPARSED_CODE':
+      break
+    default:
+      assertNever(topLevelElement)
+  }
+}
+
+type FileToUidMapping = Map<string, Set<string>>
+
+let CachedUidsPerFile = new WeakMap<ParseSuccess, GetAllUniqueUidsResultForFile>()
+
+function collectUidsForFile(
+  filePath: string,
+  parseSuccess: ParseSuccess,
+): GetAllUniqueUidsResultForFile {
+  const cachedResult = CachedUidsPerFile.get(parseSuccess)
+  if (cachedResult != null) {
+    return cachedResult
+  }
+
+  let workingUniqueIDSet = emptySet<string>()
+  let workingDupliactedUIDs: DuplicateUIDsResult = new Map()
+  fastForEach(parseSuccess.topLevelElements, (tle) => {
+    extractUIDFromTopLevelElement(workingUniqueIDSet, workingDupliactedUIDs, filePath, tle)
+  })
+  const result = { uniqueIDs: workingUniqueIDSet, duplicateIDs: workingDupliactedUIDs }
+  CachedUidsPerFile.set(parseSuccess, result)
+  return result
+}
+
+function collectUidsForEachFile(projectContents: ProjectContentTreeRoot): GetAllUniqueUIDsResult {
+  let workingFileToUidMapping: FileToUidMapping = new Map()
+  let workingDupliactedUIDs: DuplicateUIDsResult = new Map()
+  walkContentsTreeForParseSuccess(projectContents, (filePath, parseSuccess) => {
+    const mappingsForFile = collectUidsForFile(filePath, parseSuccess)
+    workingFileToUidMapping.set(filePath, mappingsForFile.uniqueIDs)
+    mappingsForFile.duplicateIDs.forEach((file, uid) => {
+      workingDupliactedUIDs.set(uid, file)
+    })
+  })
+  return {
+    filePathToUids: workingFileToUidMapping,
+    duplicateIDs: workingDupliactedUIDs,
+  }
+}
+
+export function clearCachedUidsPerFileForTests() {
+  CachedUidsPerFile = new WeakMap()
+}
+
+export function getFilePathForUid(mapping: FileToUidMapping, uid: string): string | null {
+  return mapFirstApplicable(mapping.entries(), ([filePath, uidMapping]) => {
+    if (uidMapping.has(uid)) {
+      return filePath
+    }
+    return null
+  })
+}
+
+export function getUniqueUidsMappingInner(
+  projectContents: ProjectContentTreeRoot,
+): GetAllUniqueUIDsResult {
+  return collectUidsForEachFile(projectContents)
+}
+
+export const getUidMappings = memoize(getUniqueUidsMappingInner)
+
+export function getAllUniqueUidsFromMapping(mapping: FileToUidMapping): Array<string> {
+  let result: Array<string> = []
+  for (const [filePath, uids] of mapping.entries()) {
+    result.push(...Array.from(uids))
```

Comment:
Save creating an intermediary array:
```suggestion
    uids.forEach(uid => result.push(uid))
```

Reply:
`result.push(...Array.from(uids))` is from somewhere on the internet, it is accepted as the fastest way to achieve copying. I ran a Benny benchmark and with the change the code would be somewhere between 2-6% slower